workers/unity/Packages/io.improbable.worker.sdk/EventTracing.cs

```diff
@@ -0,0 +1,349 @@
+        }
+
+        public static unsafe SpanId GetNullSpanId()
+        {
+            var nativeSpanId = new SpanId();
+            var nullSpanId = CEventTrace.SpanIdNull();
+            ApiInterop.Memcpy(nativeSpanId.Data, nullSpanId.Data, (UIntPtr) 16);
+            return nativeSpanId;
+        }
+    }
+
+    public unsafe class EventTracer : IDisposable
+    {
+        private readonly CEventTrace.EventTracer eventTracer;
+
+        public bool IsEnabled { get; private set; }
+
+        public EventTracer(EventTracerParameters[] parameters)
+        {
+            ParameterConversion.ConvertEventTracer(parameters, out var tracerParameters);
+
+            fixed (CEventTrace.EventTracerParameters* fixedParameters = tracerParameters)
+            {
+                eventTracer = CEventTrace.EventTracerCreate(fixedParameters);
+            }
+        }
+
+        public void Dispose()
+        {
+            eventTracer.Dispose();
+        }
+
+        public void Enable()
+        {
+            CEventTrace.EventTracerEnable(eventTracer);
+            IsEnabled = true;
+        }
+
+        public void Disable()
+        {
+            CEventTrace.EventTracerDisable(eventTracer);
+            IsEnabled = false;
+        }
+
+        public void SetActiveSpanId(SpanId spanId)
+        {
+            CEventTrace.EventTracerSetActiveSpanId(eventTracer, ParameterConversion.ConvertSpanId(spanId));
+        }
+
+        public void ClearActiveSpanId(SpanId spanId)
+        {
+            // Uncomment once Worker SDK updated to 14.8.0 (version where EventTracerClearActiveSpanId is added)
+            // CEventTrace.EventTracerClearActiveSpanId(eventTracer, ParameterConversion.ConvertSpanId(spanId));
+        }
+
+        public SpanId GetActiveSpanId()
+        {
+            var nativeSpanId = CEventTrace.EventTracerGetActiveSpanId(eventTracer);
+            var spanId = new SpanId();
+            ApiInterop.Memcpy(spanId.Data, nativeSpanId.Data, (UIntPtr) 16);
+            return spanId;
+        }
+
+        public void AddSpanToCauses(int causeCount, SpanId[] causes)
+        {
+            var createdSpan = CEventTrace.EventTracerAddSpan(eventTracer, null, (uint) causeCount);
+        }
+
+        public void AddEvent(Event @event)
+        {
+            CEventTrace.EventTracerAddEvent(eventTracer, ParameterConversion.ConvertEvent(@event));
+        }
+
+        public bool ShouldSampleEvent(Event @event)
+        {
+            return CEventTrace.EventTracerShouldSampleEvent(eventTracer, ParameterConversion.ConvertEvent(@event)) > 0;
+        }
+    }
+
+    public unsafe class TraceEventData : IDisposable
+    {
+        internal GCHandle eventData;
+
+        private readonly GcHandlePool fieldHandles;
+
+        public TraceEventData()
+        {
+            var internalEventData = CEventTrace.EventDataCreate();
+            eventData = GCHandle.Alloc(internalEventData, GCHandleType.Pinned);
```

Comment:
Why do we need to pin the `internalEventData` field? The location of the wrapper class in the heap doesn't matter as we care about the pointer to the native object inside of it

Reply:
So I pinned the `internalEventData` since I needed to get the `IntPtr` for the `EventData` class. The other way I could see to do it is by using the `DangerousGetHandle()` method from the `SafeHandle` class.

Reply:
Based on what Sonar cloud [complained about](https://sonarcloud.io/code?id=spatialos_gdk-for-unity&pullRequest=1452&selected=spatialos_gdk-for-unity%3APackages%2Fio.improbable.worker.sdk%2FParameterConversion.cs&line=59) it doesn't seem like I should use `DangerousGetHandle()`

Reply:
It looks like in the cases where you need access to the underlying pointer, the Worker SDK adds a method to expose it: 

https://code.i8e.io/github.com/improbable/everything@5eeb0768de55361b1977ce65b406921389f094a2/-/blob/worker_sdk/csharp_cinterop/sdk/worker/internal/Handles.cs#L105

Of course this only serves here to silence SonarCloud about `DangerousGetHandle()` 😂 

Imo - the warnings its complaining about are valid, but we can safe guard against the safety issues by ensuring that the `EventDataHandle` hasn't been disposed before calling `ConvertEvent`. 

Its also important to understand the context of `SafeHandle` which was originally designed for [Windows OS handles](https://docs.microsoft.com/en-us/dotnet/api/system.runtime.interopservices.safehandle?view=netcore-3.1#why-safehandle).

Reply:
Oh right! I did wonder why that method was included in the Worker SDK 😆 I've added it to the EventData class, I had to modify it a little compared to the Worker SDK, but it does the same thing I think.